client/my-sites/domains/domain-management/list/bulk-edit-contact-info.jsx

```diff
@@ -0,0 +1,218 @@
+	static propTypes = {
+		handleSaveContactInfo: PropTypes.func.isRequired,
+		isDisabled: PropTypes.bool,
+		isSubmitting: PropTypes.bool,
+		onTransferLockOptOutChange: PropTypes.func.isRequired,
+		emailOnly: PropTypes.bool,
+		domainNamesList: PropTypes.array,
+	};
+
+	static defaultProps = {
+		isDisabled: false,
+		isSubmitting: false,
+		emailOnly: false,
+	};
+
+	state = {
+		hasSetContactDetailsFromCache: false,
+		hasSetInitialContactEmail: false,
+		contactDetails: {},
+		errorMessages: {},
+	};
+
+	componentDidMount() {
+		if ( Object.keys( this.props.userSettings ?? {} ).length === 0 ) {
+			this.props.fetchUserSettings();
+		}
+
+		if ( Object.keys( this.props.contactDetailsCache ?? {} ).length === 0 ) {
+			this.props.requestContactDetailsCache();
+		}
+
+		this.debouncedValidateContactDetails = this.debounce( this.validateContactDetails, 500 );
+	}
+
+	componentDidUpdate() {
+		const { new_user_email, user_email } = this.props.userSettings ?? {};
+		const accountEmail = new_user_email?.length > 0 ? new_user_email : user_email;
+
+		if (
+			( accountEmail ?? '' ).length > 0 &&
+			( this.state.contactDetails?.email ?? '' ).length === 0 &&
+			false === this.state.hasSetInitialContactEmail
+		) {
+			this.setInitialContactEmail( accountEmail );
+		}
+
+		if (
+			Object.keys( this.props.contactDetailsCache ?? {} ).length > 0 &&
+			false === this.state.hasSetContactDetailsFromCache
+		) {
+			const { email, ...contactDetailsCache } = this.props.contactDetailsCache;
+			this.setContactDetailsFromCache( contactDetailsCache );
+		}
+	}
+
+	debounce = ( func, delay, { leading } = {} ) => {
+		let timerId;
+
+		return ( ...args ) => {
+			if ( ! timerId && leading ) {
+				func( ...args );
+			}
+			clearTimeout( timerId );
+
+			timerId = setTimeout( () => func( ...args ), delay );
+		};
+	};
+
+	isLoading() {
+		return (
+			Object.keys( this.props.contactDetailsCache ?? {} ).length === 0 ||
+			this.props.isRequestingContactDetailsCache ||
+			this.props.isFetchingUserSettings
+		);
+	}
+
+	setInitialContactEmail = ( email ) => {
+		this.setState( { hasSetInitialContactEmail: true }, () => {
+			this.updateDomainContactFields( { email } );
+		} );
+	};
+
+	setContactDetailsFromCache = ( data ) => {
+		this.setState( { hasSetContactDetailsFromCache: true }, () => {
+			this.updateDomainContactFields( data );
+		} );
+	};
+
+	updateDomainContactFields = ( data ) => {
+		const newContactDetails = Object.assign( {}, this.state.contactDetails, data );
+		this.setState( { contactDetails: newContactDetails } );
+		Object.keys( newContactDetails ).length > 0 &&
+			this.debouncedValidateContactDetails( newContactDetails );
+	};
+
+	onTransferLockOptOutChange = ( event ) => {
+		this.props.onTransferLockOptOutChange( event.target.checked );
+	};
+
+	validateContactDetails = ( contactDetails ) => {
+		if ( this.isLoading() ) {
+			return;
+		}
+
+		wpcom.validateDomainContactInformation(
+			contactDetails,
+			this.props.domainNamesList,
```

Comment:
`domainNamesList` is an optional prop, so this might be undefined. `validateDomainContactInformation` says it requires a `string[]` type, so we should probably have a default here just in case.

```suggestion
			this.props.domainNamesList ?? [],
```

Reply:
👍